### PR TITLE
integrate faster erasure code

### PR DESCRIPTION
**Pull request by drahnr**
Integration of https://github.com/drahnr/rs-ec-perf into polkadot.

* [x] test for correctness
* [x] decide about the code ownership - future stable code will live under https://github.com/paritytech/reed-solomon-novelpoly/ - development will continue at https://github.com/drahnr/rs-ec-perf for now due to a good enough CI already being there
* [x] improve API surface
* [x] remove dead code
* [ ] :construction: decide if we want to merge this now with a preliminary representation which will require a migration since they will have incompatible on wire representations or wait a few more days(?) until @burdges settled on the final representation?

### CODE REVIEW

File: erasure-coding/src/lib.rs [hunk 1]
```diff
@@ -75,76 +73,41 @@ pub enum Error {
...
-	})
-}
+	let n_wanted = n_validators;
+	let k_wanted = recovery_threshold(n_wanted)?;
 
-/// Obtain a threshold of chunks that should be enough to recover the data.
-pub fn recovery_threshold(n_validators: usize) -> Result<usize, Error> {
-	let n_faulty = n_faulty(n_validators)?;
+	if n_wanted > MAX_VALIDATORS as usize {
+		return Err(Error::NotEnoughValidators);
```
Comment (drahnr) [inline erasure-coding/src/lib.rs:100]:
```suggestion
		return Err(Error::TooManyValidators);
```

**Comment by rphmeier** [comment]
Let's merge this now. We can always bump the dependency crate version as long as this is before parachains go live on Kusama.

**Review by rphmeier — APPROVED** [review]

**Comment by burdges** [comment]
Are we still at 400x faster?  Or did we wind up a bit faster or slower?

**Comment by drahnr** [comment]
> Are we still at 400x faster? Or did we wind up a bit faster or slower?

That was only for the particular case of 10 MB and 2000 validators, given the current state in master it should be +-10% - running the upper bound benches again.

**Comment by drahnr** [comment]
```
novelpoly encode upper bound                                                                           
                        time:   [440.05 ms 441.33 ms 442.99 ms]
                        change: [-0.1008% +0.4453% +1.1103%] (p = 0.21 > 0.05)
                        No change in performance detected.

novelpoly decode upper bound                                                                           
                        time:   [746.45 ms 747.29 ms 748.25 ms]

Benchmarking naive encode upper bound: Warming up for 1.0000 s
Warning: Unable to complete 10 samples in 70.0s. You may wish to increase target time to 1804.6s.
naive encode upper bound                                                                          ^[[6~^[[6~
                        time:   [180.57 s 181.28 s 182.45 s]

Benchmarking naive decode upper bound: Warming up for 1.0000 s
Warning: Unable to complete 10 samples in 70.0s. You may wish to increase target time to 2028.9s.
naive decode upper bound                                                                          
                        time:   [202.66 s 202.76 s 202.87 s]

```

**Comment by drahnr** [comment]
For sake of completeness (latest fuzz run):
```
------------------------[  0 days 02 hrs 04 mins 10 secs ]----------------------
  Iterations : 18,820,855 [18.82M]
  Mode [3/3] : Feedback Driven Mode
      Target : hfuzz_target/x86_64-unknown-linux-gnu/release/fuzzit
     Threads : 8, CPUs: 16, CPU%: 1600% [100%/CPU]
       Speed : 2,070/sec [avg: 2,526]
     Crashes : 0 [unique: 0, blocklist: 0, verified: 0]
    Timeouts : 0 [1 sec]
 Corpus Size : 235, max: 8,192 bytes, init: 594 files
  Cov Update : 0 days 02 hrs 04 mins 00 secs ago
    Coverage : edge: 652/3,293 [19%] pc: 1 cmp: 25,738
---------------------------------- [ LOGS ] ------------------/ honggfuzz 2.4 /-
```

**Pull request merged**